cmd/ecc/join.go

```diff
@@ -0,0 +1,330 @@
+		}
+	}
+	return &JoinedReaders{
+		src:    src,
+		err:    errors,
+		offset: offset,
+	}
+}
+
+// JoinedReaders groups a set of data sources
+// and manages reading from them concurrently.
+type JoinedReaders struct {
+	src    []io.ReaderAt
+	err    []error
+	offset int64
+
+	rErr error
+}
+
+// Read tries to read just as many shards
+// as necessary into the buffer such that the
+// erasure coding can reconstruct missing data
+// shards, if any.
+//
+// Let d be the number of data shards. Read spawns
+// d go routines - each tries to read one complete
+// shard from its data source. Whenever a go routine
+// completes but returns an error this data source
+// is considered unavailable and another go routine
+// is spawned. This new go routine tries to read from
+// the next data source. If no more data sources are
+// available it returns ErrNoReadQuorum.
+//
+// Read minimizes the total number of reads / connections
+// and tries to read as many data shards before
+// it tries to fetch any parity shards. Minimizing
+// the number of reads / connections is important to
+// avoid any unnecessary I/O or network load. Also, if
+// a data shard is available we should prefer fetching it
+// instead of reading a parity shard to avoid unnecessary
+// erasure coding reconstruction.
+func (r *JoinedReaders) Read(b *Buffer) error {
+	if len(r.src) != len(b.shards) {
+		panic("ecc: number of data sources does not match the number of shards")
+	}
+	if r.rErr != nil {
+		return r.rErr
+	}
+
+	b.Reset()
+	results, next, err := r.spawn(b)
+	if err != nil {
+		r.rErr = err
+		return r.rErr
+	}
+
+	// Mark all shards as missing.
+	// When a read from a data source succeeds
+	// the particular shard will get marked
+	// as available again.
+	for i := range b.shards {
+		b.shards[i] = b.shards[i][:0]
+	}
+
+	return r.join(b, next, results)
+}
+
+type readResult struct {
+	ID  int   // The ID / index of the shard / data source
+	N   int   // The number of bytes read from the data source
+	Err error // Any error that happened during reading
+}
+
+// spawn creates just as many concurrent go routines as necessary to read
+// enough shards to reconstruct the content data. If all data shards are
+// available each go routine tries to read a data shard.
+// It returns a channel which hold the read results once they are available.
+// The returned next index points to the next data source from which a go
+// routine should read if any of the spawned reads fails.
+func (r *JoinedReaders) spawn(b *Buffer) (results chan readResult, next int, err error) {
+	results = make(chan readResult, len(r.src))
+	var spawned int
+	for spawned < len(b.data) && next < len(b.shards) {
+		if r.err[next] == nil {
+			go readFrom(r.src[next], next, r.offset, b.shards[next], results)
+			spawned++
+		}
+		next++
+	}
+	// If we could not spwan a single go routine we should return an error
+	// (or panic) since a join would block forever since the results channel
```

Comment:
This code relies on the invariant that `spawn` is only called if there are enough data sources available to achieve read quorum. More precisely, the caller should never call this function if no data source is available - since there never was read quorum.
This `err = ErrReadQuorum` is actually redoing the job of the caller.